weather_dl/download_pipeline/parsers.py

```diff
@@ -69,6 +69,86 @@ def date(candidate: str) -> datetime.date:
     return converted
 
 
+def time(candidate: str) -> datetime.time:
+    """Converts ECMWF-format time strings into a `datetime.time`.
+
+    Accepted time formats:
+    - HH:MM
+    - HHMM
+    - HH
+
+    For example:
+    - 18:00
+    - 1820
+    - 18
+
+    Note: If MM is omitted it defaults to 00.
+    """
+    converted = None
+
+    accepted_formats = ["%H", "%H:%M", "%H%M"]
+
+    for fmt in accepted_formats:
+        try:
+            converted = datetime.datetime.strptime(candidate, fmt).time()
+            break
+        except ValueError:
+            pass
+
+    if converted is None:
+        raise ValueError(
+            f"Not a valid time: '{candidate}'. Please use valid format."
+        )
+
+    return converted
+
+
+def day_month_year(key: str, candidate: t.Any) -> int:
+    """Converts day, month and year strings into 'int'.
+
+    Also validates the day and month input.
+    """
+    converted = None
+
+    def validate(key: str, candidate: int) -> None:
+        nonlocal converted
+        try:
+            if key == "day":
+                assert 1 <= candidate <= 31
+            if key == "month":
+                assert 1 <= candidate <= 12
```

Comment:
Potentially, you could pass specific error message to the user with this assert message. If you want to present it as a top level `ValueError`, consider using the `raise from` syntax. 

https://docs.python.org/3/tutorial/errors.html#exception-chaining

Reply:
In addition to adding an assert message, you'd need to refactor this function.

Reply:
Refactored the function, please have a look and provide your feedback.